Middleware docs (#5625)
vuestorefront/vue-storefront: packages/core/docs/advanced/calling-platform-api.md

# Calling integration APIs

## How do Integrations work?

> In Vue Storefront, single integration has is own sort of SDK library (named API-client), that contains set of functions. Each function is dedicated to one action, endpoint or a feature eg. `getProduct`, `loadCart` `addToCart`. Therefore, when you call for instance, `getProduct`, it will request the integration API and return its response in the original format of given platform.

Reviewer: ```suggestion
In Vue Storefront, single integration has its own sort of software development kit (SDK) library. This library is named API-client and contains a set of functions. Each function is dedicated to one action, endpoint or a feature eg. `getProduct`, `loadCart` `addToCart`. For example, when you call `getProduct`, it will request the integration API and return its response in the original format of a given platform.
```

## Accessing integration methods on the frontend

> To access API-client functions, you can use the composable function `useVSFContext` that reads from the application context every possible integration so you can easily access it.

Reviewer: **suggestion**: I believe the function name is explanatory here and we can shorten this sentence:

```suggestion
To access API-client functions, you can use the composable function `useVSFContext`:
```

```ts
// each platform has different tag to access its methods, eg $spryker, $storyblok etc.
const { $ct } = useVSFContext();

$ct.api.getProduct({ id: 1 })
```

In the example above we have accessed the API-client for commercetools (Each integration has dedicated tag name, look at [context docs](/advanced/context)) and we called function `getProduct`.

## Extending Integrations

> Sometimes, it's necessary to override the original behavior either API-client or even an entire request that comes from an external platform.

Reviewer: ```suggestion
Sometimes, it's necessary to override the original behavior for either API-client or even an entire request that comes from an external platform.
```

> # Calling platform API

Author: is it really "integration APIs"? using API client you are calling external API / platform API

> The Vue Storefront has its own way to communicate with other platform APIs. First of all, each integration implements an API-client that defines an interaction and connection with a given platform. Each time you make a call to the external platform, you actually use API-client beneath. We use that in all of the integrations and also you can reach the API-client in your project if it's needed.

Reviewer 2: ```suggestion
In Vue Storefront each integration implements an API-client that defines an interaction and connection with a given platform. Each time you make a call to the external platform, it's handled by API-client underneath. We use that in all of the integrations. If needed, you can reach the API-client in your project too.
```

Reviewer 2: This is a little shorter and in my opinion easier to understand

> The Vue Storefront has its own way to communicate with other platform APIs. First of all, each integration implements an API-client that defines an interaction and connection with a given platform. Each time you make a call to the external platform, you actually use API-client beneath. We use that in all of the integrations and also you can reach the API-client in your project if it's needed.

Reviewer 3: 1. This paragraph provides 0 knowledge about API client works and what it is
2. Does user even need this information?

> The Vue Storefront has its own way to communicate with other platform APIs. First of all, each integration implements an API-client that defines an interaction and connection with a given platform. Each time you make a call to the external platform, you actually use API-client beneath. We use that in all of the integrations and also you can reach the API-client in your project if it's needed.

Reviewer 3: also you're using API CLIENT 3x here but explain it later ;)

> ## What is an integration API Client?

Author: but integrations are also composables or even UI, in this section, we are talking about the only middleware/api-client

> Basically, API-client is a sort of SDK for a given platform you integrate with. It has a set of functions, each is dedicated to one action, endpoint, feature, or something else. It's aiming to provide an abstraction for the developers when it comes to making calls.

Reviewer 2: ```suggestion
API-client is a sort of an SDK for a platform you integrate with. It has a set of functions, each dedicated to one action, endpoint, feature, etc. It's aiming to provide an abstraction for the developers when it comes to making calls.
```

> Basically, API-client is a sort of SDK for a given platform you integrate with. It has a set of functions, each is dedicated to one action, endpoint, feature, or something else. It's aiming to provide an abstraction for the developers when it comes to making calls.

Reviewer 3: ```suggestion
Integration API-client is a sort of SDK for a given platform you integrate with. It has a set of functions, each is dedicated to one action, endpoint or a feature eg. `getProduct` , `loadCart` `addToCart`
```

## Using useVSFContext to access integration API Client methods

To access API-client functions, you can use the composable function `useVSFContext` that reads from the application context every possible integration so you can easily access it.

```ts

> const { $ct } = useVSFContext();

Reviewer 3: ```suggestion
// each platform has different tag to access its methods, eg $spryker, $storyblok etc.
const { $ct } = useVSFContext();
```

$ct.api.getProduct({ id: 1 })
```

> In the example above we have accessed the API-client for commercetools (Each integration has dedicated tag name, look at [context docs](/advanced/context)) and we called function `getProduct`.

Reviewer 2: ```suggestion
In the example above we access the API-client for `commercetools` and call `getProduct` function. Each integration has a dedicated tag name - for more information see [context docs](/advanced/context).
```

::: tip
Remember that API-client you are accessing from the front-end side is only a stub. It's making a call to our middleware where the real API-client is being called. For more info, jump to the [server middleware](/advanced/server-middleware) section.
:::

## Extending API Client

> Sometimes, it's necessary to override the original behavior either API-client or even an entire request that comes from an external platform. The Vue Storefront also provides possibility to do this by using [middleware extensions](/advanced/server-middleware)

Reviewer 2: ```suggestion
Sometimes, it's necessary to override the original behavior in API-client or even an entire request that comes from an external platform. The Vue Storefront also provides the possibility to do this using [middleware extensions](/advanced/server-middleware).
```

> Sometimes, it's necessary to override the original behavior either API-client or even an entire request that comes from an external platform. The Vue Storefront also provides possibility to do this by using [middleware extensions](/advanced/server-middleware)

Reviewer 3: ```suggestion
Sometimes, it's necessary to override the original behavior either API-client or even an entire request that comes from an external platform. The Vue Storefront also provides possibility to do this by using [middleware extensions](/advanced/server-middleware)
```
Instead of this please think about some use case and show the functionality as this use case
